irksome/getForm.py

```diff
@@ -124,60 +142,81 @@ def getForm(F, butch, t, dt, u0, bcs=None, bc_type="DAE"):
          either `f.interpolate(expr-c*u0)` or `f.project(expr-c*u0)`, depending
          on whether the function space for f supports interpolation or
          not.
-
     """
     v = F.arguments()[0]
     V = v.function_space()
     assert V == u0.function_space()
 
-    A = numpy.array([[Constant(aa) for aa in arow] for arow in butch.A],
-                    dtype=object)
     c = numpy.array([Constant(ci) for ci in butch.c],
                     dtype=object)
 
-    num_stages = len(c)
+    bA1, bA2 = splitting(butch.A)
+
+    try:
+        bA1inv = numpy.linalg.inv(bA1)
+    except numpy.linalg.LinAlgError:
+        bA1inv = None
+    try:
+        bA2inv = numpy.linalg.inv(bA2)
+    except numpy.linalg.LinAlgError:
+        bA2inv = None
+
+    A1 = numpy.array([[Constant(aa) for aa in arow] for arow in bA1],
+                     dtype=object)
+    if bA1inv is not None:
+        A1inv = numpy.array([[ConstantOrZero(aa) for aa in arow] for arow in bA1inv],
+                            dtype=object)
+    else:
+        A1inv = None
+    A2inv = numpy.array([[ConstantOrZero(aa) for aa in arow] for arow in bA2inv],
+                        dtype=object)
+
+    num_stages = butch.num_stages
     num_fields = len(V)
 
     Vbig = reduce(mul, (V for _ in range(num_stages)))
     # Silence a warning about transfer managers when we
     # coarsen coefficients in V
     push_parent(V.dm, Vbig.dm)
+
     vnew = TestFunction(Vbig)
-    k = Function(Vbig)
+    w = Function(Vbig)
+
     if len(V) == 1:
         u0bits = [u0]
         vbits = [v]
         if num_stages == 1:
             vbigbits = [vnew]
-            kbits = [k]
+            wbits = [w]
         else:
             vbigbits = split(vnew)
-            kbits = split(k)
+            wbits = split(w)
     else:
         u0bits = split(u0)
         vbits = split(v)
         vbigbits = split(vnew)
-        kbits = split(k)
+        wbits = split(w)
 
-    kbits_np = numpy.zeros((num_stages, num_fields), dtype=object)
+    wbits_np = numpy.zeros((num_stages, num_fields), dtype=object)
 
     for i in range(num_stages):
         for j in range(num_fields):
-            kbits_np[i, j] = kbits[i*num_fields+j]
+            wbits_np[i, j] = wbits[i*num_fields+j]
 
-    Ak = A @ kbits_np
+    A1w = A1 @ wbits_np
+    A2invw = A2inv @ wbits_np
```

Comment:
You always need `A2inv`, but sometimes it isn't available, I suspect you want to error sooner.

Reply:
Yes, it can be done when failing to form `bA2inv`